Source/Core/Core/HW/DVD/AMMediaboard.cpp

```diff
@@ -0,0 +1,1855 @@
+                       timeoutB, timeoutC, ret);
+
+        s_media_buffer[1] = s_media_buffer[8];
+        media_buffer_32[1] = ret;
+        break;
+      }
+      case AMMBCommand::GetParambyDHCPExec:
+      {
+        u32 value = media_buffer_32[2];
+
+        NOTICE_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: GetParambyDHCPExec({})\n", value);
+
+        s_media_buffer[1] = 0;
+        media_buffer_32[1] = 0;
+        break;
+      }
+      case AMMBCommand::ModifyMyIPaddr:
+      {
+        u32 NetBufferOffset = *(u32*)(s_media_buffer + 8) - NetworkCommandAddress2;
+
+        char* IP = (char*)(s_network_command_buffer + NetBufferOffset);
+
+        NOTICE_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: modifyMyIPaddr({})\n", IP);
+        break;
+      }
+      case AMMBCommand::GetLastError:
+      {
+        u32 fd = s_sockets[SocketCheck(media_buffer_32[2])];
+
+        NOTICE_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: GetLastError( {}({}) ):{}\n", fd,
+                       media_buffer_32[2], s_last_error);
+
+        s_media_buffer[1] = s_media_buffer[8];
+        media_buffer_32[1] = s_last_error;
+      }
+      break;
+      case AMMBCommand::InitLink:
+        NOTICE_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: InitLink");
+        break;
+      default:
+        ERROR_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: Command:{:03X}", *(u16*)(s_media_buffer + 2));
+        ERROR_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: Command Unhandled!");
+        break;
+      }
+
+      s_media_buffer[3] |= 0x80;  // Command complete flag
+
+      memset(memory.GetSpanForAddress(address).data(), 0, length);
+
+      ExpansionInterface::GenerateInterrupt(0x10);
+      return 0;
+    }
+
+    if (offset >= DIMMCommandVersion2_2 && offset <= 0x89000200)
+    {
+      u32 dimmoffset = offset - DIMMCommandVersion2_2;
+      memcpy(memory.GetSpanForAddress(address).data(), s_media_buffer + dimmoffset, length);
+
+      INFO_LOG_FMT(DVDINTERFACE_AMMB, "GC-AM: Read MEDIA BOARD COMM AREA (3) ({:08x})", dimmoffset);
+      PrintMBBuffer(address, length);
+      return 0;
+    }
+
+    // DIMM memory (8MB)
+    if (offset >= DIMMMemory2 && offset <= 0xFF800000)
+    {
+      u32 dimmoffset = offset - DIMMMemory2;
+      s_dimm->Seek(dimmoffset, File::SeekOrigin::Begin);
+      s_dimm->ReadBytes(memory.GetSpanForAddress(address).data(), length);
+      return 0;
+    }
+
+    if (offset == NetworkControl && length == 0x20)
+    {
+      s_netctrl->Seek(0, File::SeekOrigin::Begin);
+      s_netctrl->ReadBytes(memory.GetSpanForAddress(address).data(), length);
+      return 0;
+    }
+
+    // Max GC disc offset
+    if (offset >= 0x57058000)
+    {
+      PanicAlertFmtT("Unhandled Media Board Read:{0:08x}", offset);
+      return 0;
+    }
+
+    if (s_firmwaremap)
+    {
+      if (s_segaboot)
+      {
+        DICMDBUF[1] &= ~0x00100000;
+        DICMDBUF[1] -= 0x20;
+      }
+      memcpy(memory.GetSpanForAddress(address).data(), s_firmware + offset, length);
+      return 0;
+    }
```

Comment:
This looks incorrect based on the above comment. I think it needs to be:

```suggestion
    if (s_firmwaremap)
    {
      if (s_segaboot)
      {
        offset &= ~0x00100000;
        offset -= 0x20;
      }
      memcpy(memory.GetSpanForAddress(address).data(), s_firmware + offset, length);
      return 0;
    }
```